Jamfile

```diff
@@ -56,15 +56,52 @@ rule linking ( properties * )
 {
 	local result ;
 
+	# ssl=openssl => crypto=libcrypto
+	# this match the previous behavior
+	if <ssl>openssl in $(properties)
+		&& ! <crypto>libcrypto in $(properties)
+	{
+		EXIT "with ssl=openssl the value for crypto must be libcrypto" ;
+	}
+
+	# crypto=commoncrypto => target-os=darwin
+	if <crypto>commoncrypto in $(properties)
+		&& ! <target-os>darwin in $(properties)
+	{
+		EXIT "value of commoncrypto for crypto option is only possible in macosx" ;
+	}
+
 	# openssl libraries, if enabled
-	if <crypto>openssl in $(properties)
-		|| <crypto>openssl-sha1 in $(properties)
+	if <ssl>openssl in $(properties)
+	{
+		# exclude gcc from a regular windows build to make mingw
+		# link against the regular unix library name
+		if <target-os>windows in $(properties)
+			&& ! <toolset>gcc in $(properties)
+		{
+			result += <library>ssleay32
+				<library>libeay32
+				<library>advapi32
+				<library>user32
+				<library>shell32
+				<library>gdi32
+			;
+		}
+		else
+		{
+			result += <library>crypto <library>ssl ;
+		}
+	}
+
+	if <crypto>libcrypto in $(properties)
+		&& ! <ssl>openssl in $(properties)
```

Comment:
I don't think you need to care about the <ssl> feature here. if you make them independent.